### ISSUE TITLE
Manual playback time change through ControlBar doesn't work in Chrome

### ISSUE THREAD

**Issue by dusandjovanovic** [video-react/video-react#234]
**Description**
This bug when streaming or playing a local video happens in latest version of **Google Chrome only**. Including a ControlBar component in a video and trying to change playback time always resets the video to 00:00. This happens for both development and production builds.

1. Include ControlBar inside Player component and try to change playback time to any time offset
2. Clicking or dragging the control bar's circle in order to jump to a time offset
3. Video **playback starts from the beginning**

**Expected behavior**
Time offset follows the circle.

**Desktop**
 - OS: Windows 10 build 1809
 - Browser: Google Chrome
 - Browser version: 72.0.3626.121
 - **React version: 16.8.3**
 - video-react version: 0.10.7 latest

_I haven't tested with older React and ReactDOM versions. Could possibly be source of the problem._

**Context**
This behavior is not present in latest version of Firefox. Video's time offset is appropriate to the interaction of the control bar.

**Comment by mondaychen** [comment]
I'm not able to reproduce it on https://video-react.js.org/components/control-bar/

Can you provide a demo?

**Comment by mondaychen** [comment]
If you believe it might relate to react version, I tested it with react 16.8.3 here and it works for me https://codesandbox.io/s/6w479621x3

**Comment by dusandjovanovic** [comment]
After testing I can say that this happens in some rare use-cases when Player is deeply nested and controlled with a ref, I will provide you with the code tomorrow.

**Comment by dusandjovanovic** [comment]
Actually, no need to wait for tomorrow, here's the code:

Wrapper for react-video:
`import React from "react";
import Player from "video-react/lib/components/Player";
import ControlBar from "video-react/lib/components/control-bar/ControlBar";
import PropTypes from "prop-types";

import "video-react/dist/video-react.css";

class Video extends React.PureComponent {
    state = {
        player: null,
        source: this.props.src,
        currentTime: 0
    };

    componentDidMount() {
        if (this.props.handleStateChange)
            this.props.innerRef.current.subscribeToStateChange(this.handleStateChange);
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
        if (this.props.src !== prevProps.src)
            this.changeSource(this.props.src);
    }

    handleStateChange = state => {
        this.setState(
            {
                player: state,
                currentTime: state.currentTime
            },
            () => {
                this.props.handleStateChange(state);
            }
        );
    };

    changeSource = name => {
        this.setState(
            {
                source: name,
                currentTime: 0
            },
            () => this.props.innerRef.current.load()
        );
    };

    render() {
        return (
            <Player
                ref={this.props.innerRef}
                width={this.props.width ? this.props.width : undefined}
                poster={this.props.poster ? this.props.poster : undefined}
                autoPlay={this.props.autoPlay}
            >
                <source src={this.state.source}/>
                <ControlBar disableCompletely={this.props.disableControlBar}/>
            </Player>
        );
    }
}

Video.propTypes = {
    src: PropTypes.string.isRequired,
    handleStateChange: PropTypes.func,
    width: PropTypes.number,
    autoPlay: PropTypes.bool,
    poster: PropTypes.string,
    disableControlBar: PropTypes.bool,
    innerRef: PropTypes.any.isRequired
};

export default React.forwardRef((props, ref) => (
    <Video innerRef={ref} {...props} />
));`

And here is usage.

`<Video
          handleStateChange={props.handleStateChange}
          src={backendRoutePlayback(props.video)}
/>`

In this case changing time offset works **only once**, afterwards the Player is frozen. Only in Chrome.

**Comment by mondaychen** [comment]
I have two questions:

1. why do you need to use `forwardRef`?
2. when using this component you built, why do you pass in `innerRef` instead of `ref`?

**Comment by dusandjovanovic** [comment]
I'm forwarding a reference from a container in which the player is in, in order to completely control the Player as in changing the currentTime from a component above.

I tried "dumbing-down" my wrapper component and removing ref forwarding, `subscribeToStateChange`, `changeSource ` and unnecessary props on the Player and ControlBar but I end up getting the same problems (freezing, not changing time offset) in Chrome only.

**Comment by dusandjovanovic** [comment]
P.S. I tried using a Player w/ ControlBar **without a wrapper component** like above, and it worked fine. But in order to change the time-offset from a container upper in a tree I need to use `React.forwardRef ` or something similar, with a wrapper of course.

**Comment by mondaychen** [comment]
Will you be able to fork my demo in codebox and create a simple demo with this problem? I have a few assumptions that may cause your problem, but without a demo I'm not sure which would be the right direction

**Comment by dusandjovanovic** [comment]
Could you provide me with the link to the demo's repository? In the meantime I'll provide you with my complete code resulting faulty behavior. In this case video's time offset is not programmatically changed, there is another context with those requests.

VideoForm component (choosing a video file from the file-system):

```
import React from "react";
import PropTypes from "prop-types";

class FormVideo extends React.Component {
    state = {
        file:
            this.props.formDefaults && this.props.formDefaults.file
                ? this.props.formDefaults.file
                : null,
        fileURL:
            this.props.formDefaults && this.props.formDefaults.fileUrl
                ? this.props.formDefaults.fileUrl
                : null
    };

    handleSelectedFile = event => {
        const file = event.target.files[0];
        const reader = new FileReader();
        reader.readAsDataURL(file);
        reader.onloadend = () => {
            this.setState(() => ({
                file: file,
                fileURL: reader.result
            }));
            this.props.submitHandler(file, reader.result);
        };
        reader.onerror = () => {
            this.props.errorHandler({
                hasError: true,
                description: "There was an error adding a video, please try again."
            });
        };
    };

    render() {
        return (
            <React.Fragment>
                {this.props.children({
                    ...this.state,
                    handleSelectedFile: this.handleSelectedFile
                })}
            </React.Fragment>
        );
    }
}

FormVideo.propTypes = {
    formDefaults: PropTypes.object,
    submitHandler: PropTypes.func.isRequired,
    errorHandler: PropTypes.func.isRequired,
    children: PropTypes.func.isRequired
};

export default FormVideo;
```


Wrapper component for react-video:

```
import React from "react";
import Player from "video-react/lib/components/Player";
import ControlBar from "video-react/lib/components/control-bar/ControlBar";
import PropTypes from "prop-types";

import "video-react/dist/video-react.css";

class Video extends React.Component {
    state = {
        player: null,
        source: this.props.src,
        currentTime: 0
    };

    componentDidMount() {
        if (this.props.handleStateChange)
            this.props.innerRef.current.subscribeToStateChange(this.handleStateChange);
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
        if (this.props.src !== prevProps.src)
            this.changeSource(this.props.src);
    }

    handleStateChange = state => {
        this.setState(
            {
                player: state,
                currentTime: state.currentTime
            },
            () => {
                this.props.handleStateChange(state);
            }
        );
    };

    changeSource = name => {
        this.setState(
            {
                source: name,
                currentTime: 0
            },
            () => this.props.innerRef.current.load()
        );
    };

    render() {
        return (
            <Player
                ref={this.props.innerRef}
                width={this.props.width ? this.props.width : undefined}
                poster={this.props.poster ? this.props.poster : undefined}
                autoPlay={this.props.autoPlay}
            >
                <source src={this.state.source}/>
                <ControlBar disableCompletely={this.props.disableControlBar}/>
            </Player>
        );
    }
}

Video.propTypes = {
    src: PropTypes.string.isRequired,
    handleStateChange: PropTypes.func,
    width: PropTypes.number,
    autoPlay: PropTypes.bool,
    poster: PropTypes.string,
    disableControlBar: PropTypes.bool,
    innerRef: PropTypes.any.isRequired
};

export default React.forwardRef((props, ref) => (
    <Video innerRef={ref} {...props} />
));
```

Container for both (stripped of some code that's not relevant):

```
import React from "react";
import Video from "../../../../components/logic/video/video";
import FormVideo from "../../../../components/logic/form-video/formVideo";
import {formElements} from "../../../../assets/constants/containers/videos/videoForm";
import Button from "@material-ui/core/Button";
import PropTypes from "prop-types";

import withErrorHandler from "../../../../hoc/withErrorHandler/withErrorHandler";

import {styles} from "./stylesheet";
import withStyles from "@material-ui/core/styles/withStyles";

class VideoForm extends React.Component {
    constructor(props) {
        super(props);
        this.innerVideoRef = React.createRef();
    }

    state = {
        error: {
            hasError: false,
            name: null,
            description: null
        }
    };

    errorHandler = value => {
        this.setState(() => ({
            error: {
                ...value
            }
        }));
    };

    submitHandler = data => {
        this.props.action(
            data.title.value,
            data.description.value,
            data.price.value,
            data.source.value,
            data.source.valueUrl,
            this.innerVideoRef.current.getState().player.duration
        );
    };

    render() {
        const {classes} = this.props;

        return (
            <div className={classes.content}>
                        // more code here..
                        <React.Fragment>
                                <FormVideo
                                    formDefaults={{
                                        file: this.props.values.source,
                                        fileUrl: this.props.values.sourceUrl
                                    }}
                                    submitHandler={(file, fileUrl) =>
                                        mediaChangedHandler(file, fileUrl, inputs.source.id)
                                    }
                                    errorHandler={this.errorHandler}
                                >
                                    {({fileURL, handleSelectedFile}) => (
                                        <div className={classes.imgInputs}>
                                            {fileURL ? (
                                                <Video
                                                    handleStateChange={null}
                                                    innerRef={this.innerVideoRef}
                                                    src={fileURL}
                                                />
                                            ) : null}

                                            <input
                                                onChange={handleSelectedFile}
                                                accept="video/*"
                                                className={classes.input}
                                                id="flat-button-file"
                                                multiple
                                                type="file"
                                            />

                                            <label htmlFor="flat-button-file">
                                                <Button
                                                    component="span"
                                                    variant="outlined"
                                                    color="primary"
                                                    className={classes.button}
                                                >
                                                    Upload a video
                                                </Button>
                                            </label>
                                        </div>
                                    )}
                                </FormVideo>
                                // more code here..
                        </React.Fragment>
            </div>
        );
    }
}

VideoForm.propTypes = {
    classes: PropTypes.object.isRequired,
    values: PropTypes.object.isRequired,
    handleBack: PropTypes.func.isRequired,
    handleNext: PropTypes.func.isRequired,
    handleSkip: PropTypes.func.isRequired,
    activeStep: PropTypes.number.isRequired,
    steps: PropTypes.array.isRequired,
    action: PropTypes.func.isRequired
};

export default withStyles(styles)(withErrorHandler(VideoForm));
```

**Comment by dusandjovanovic** [comment]
A quick update. With the latest version you committed the upper mentioned problem of freezing is solved.

The problem with timeOffset re-initiating to 0s when changed from the ControlBar remains. Here is the code. Here the ref is not forwarded and is kept locally.

```
import React from "react";
import Player from "video-react/lib/components/Player";
import ControlBar from "video-react/lib/components/control-bar/ControlBar";
import PropTypes from "prop-types";

import "video-react/dist/video-react.css";

class Video extends React.Component {
    constructor(props) {
        super(props);
        this.videoRef = React.createRef();
        this.state = {
            source: props.src,
            currentTime: 0
        };
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
        if (this.state.source !== this.props.src)
            this.changeSource(this.props.src);
    }

    componentDidMount() {
        if (this.props.handleStateChange)
            this.videoRef.current.subscribeToStateChange(this.handleStateChange);
    }

    handleStateChange = state => {
        this.setState(
            {
                currentTime: state.currentTime
            },
            () => {
                this.props.handleStateChange(state);
            }
        );
    };

    changeSource = name => {
        this.setState(
            {
                source: name,
                currentTime: 0
            },
            () => this.videoRef.current.load()
        );
    };

    render() {
        return (
            <Player
                ref={this.videoRef}
                width={this.props.width ? this.props.width : undefined}
                poster={this.props.poster ? this.props.poster : undefined}
                autoPlay={this.props.autoPlay}
            >
                <source src={this.state.source}/>
                <ControlBar disableCompletely={this.props.disableControlBar}/>
            </Player>
        );
    }
}

Video.propTypes = {
    src: PropTypes.string.isRequired,
    handleStateChange: PropTypes.func,
    width: PropTypes.number,
    autoPlay: PropTypes.bool,
    poster: PropTypes.string,
    disableControlBar: PropTypes.bool
};

export default Video;

```

And is used in a more simple way, without management of input since the video src comes from the backend and is streamed. Again, stripped of unnecessary code.

```
import React from "react";
import Video from "../../../../../components/logic/video-unmanaged/videoUnmanaged";
import PropTypes from "prop-types";

import {backendRoutePlayback} from "../../../../../utils/constants/api-constants";

import {styles} from "./stylesheet";
import withStyles from "@material-ui/core/styles/withStyles";

const workoutContent = props => {
    const {classes} = props;

    return (
        <div className={classes.root}>
            <div className={classes.video}>
                <Video
                    handleStateChange={props.handleStateChange}
                    src={backendRoutePlayback(props.video)}
                />
            </div>
            // more code here, relying on timeOffset for rendering
           // timeOffset is used from the Video component and updates the state upper in the tree
           // props.handleStateChange is passed to a Video component and should follow its offset
        </div>
    );
};

workoutContent.propTypes = {
    classes: PropTypes.object.isRequired,
    video: PropTypes.string.isRequired,
    steps: PropTypes.arrayOf(PropTypes.object),
    stepActive: PropTypes.number,
    stepPercentage: PropTypes.number,
    handleStateChange: PropTypes.func.isRequired
};

export default withStyles(styles)(React.memo(workoutContent));
```

To sum things up, Video wrapper should provide an api for following video's time offset and change state in the parent accordingly, with handleStateChange({timeOffset: seconds}).

**Comment by mondaychen** [comment]
Go to https://codesandbox.io/s/6w479621x3 and click fork. After you are done, paste the url back here. It's the easiest way I know to create a demo.

**Comment by dusandjovanovic** [comment]
https://codesandbox.io/s/qx0q5rljr6 

I can't seem to reproduce the same issue right now, at least not with the stripped-down version of what I have in my project. I will add more code to the same fork later today or tomorrow and will let you know when I do.

**Comment by dusandjovanovic** [comment]
I tried adding all of my code to the fork with mock data but that would take too much time.

In short, what I'm doing, is relying on handleStateChange to update the state of the parent (unlike just logging to the console as I did in the fork). That state is used to render React elements next to the video.

Here is the screenshot: https://ibb.co/9Y9MrxK

The screenshot is taken in Firefox, where the video is streaming from a back-end (http atm). In Firefox, clicking or dragging the circle on the ControlBar works as it should, changing the time offset to the end of the downloaded portion of the video. In Chrome the same action just resets the video to 0:00.

**Comment by dusandjovanovic** [comment]
Here is an update. I'll provide you with two screenshots containing logging for react-video's internal state accessed through `subscribeToStateChange`. The first screenshot is in Chrome, the second in Firefox, both latest versions (stable).

```
componentDidMount() {
        if (this.props.handleStateChange)
            this.videoRef.current.subscribeToStateChange(this.handleStateChange);
    }

    handleStateChange = videoState => {
        console.log(videoState);
        this.setState(
            {
                currentTime: videoState.currentTime
            },
            () => {
                this.props.handleStateChange(videoState);
            }
        );
    };
```

https://ibb.co/1fJwBbk
https://ibb.co/GnwrzmB

You can see the state changing when there is a change in `seekingTime`. This happens when the video is streamed, not played locally.

**Comment by dusandjovanovic** [comment]
If you need complete logs from the console just let me know.

**Comment by mondaychen** [comment]
So it's no longer an issue related with `ref`, correct?

I'm willing to help but it's very hard for me when I don't have a demo that I can reproduce the issue on my end.

**Comment by dusandjovanovic** [comment]
It's not reference related. I found a problem, it's present event when using a default HTML5 video.

Hosting a video on a php server and streaming it doesn't work as it should be in Chrome. Here is a quote I found regarding this issue.

> Chrome makes multiple HTTP request with Range header for media file and expects HTTP Content-Range and Content-Length response header with 206 Partial Content status (read this). If server send '200 OK' status Chrome will not play the media file, but Firefox will. If server send '200 OK' status with response header 'Accept-Ranges: none' then Chrome will play media file, but jumping will not work. I assume you have implemented hook_file_download() already. That hook should return array of appropriate HTTP headers. It's enough to return Accept-Ranges header > 

You can read more about this [here](https://drupal.stackexchange.com/questions/234721/unable-to-jump-to-a-specific-spot-seek-when-mp4-or-mp3-is-in-private-field).

**Comment by mondaychen** [comment]
Ah, so the problem should be fixable on backend?

**Comment by dusandjovanovic** [comment]
I suppose, I'll see with my coworker who's in charge of that.

**Comment by dusandjovanovic** [comment]
Can confirm that the issue was backend related. Closing it.
